In [2]:
%load_ext sql

In [3]:
%sql postgresql://appdev@data/appdev

'Connected: appdev@appdev'

# Tables and columns in the f1db schema

In [ ]:
%sql SELECT * from information_schema.tables WHERE table_schema = 'f1db';

In [ ]:
%sql SELECT * from information_schema.columns WHERE table_schema = 'f1db';

## Using a single join and a grouping, write a query that allows you to find the forename, surname, driverid and the total number of wins for each driver in the f1db schema.

In [15]:
%%sql
SELECT drivers.forename, drivers.surname, drivers.driverid, count(*) as "wins"
FROM drivers
INNER JOIN results ON
results.driverid = drivers.driverid
WHERE results.position = 1
GROUP BY drivers.driverid
ORDER BY count(*) DESC

107 rows affected.


[('Michael', 'Schumacher', 30, 91),
 ('Lewis', 'Hamilton', 1, 60),
 ('Alain', 'Prost', 117, 51),
 ('Sebastian', 'Vettel', 20, 46),
 ('Ayrton', 'Senna', 102, 41),
 ('Fernando', 'Alonso', 4, 32),
 ('Nigel', 'Mansell', 95, 31),
 ('Jackie', 'Stewart', 328, 27),
 ('Jim', 'Clark', 373, 25),
 ('Niki', 'Lauda', 182, 25),
 ('Juan', 'Fangio', 579, 24),
 ('Nico', 'Rosberg', 3, 23),
 ('Nelson', 'Piquet', 137, 23),
 ('Damon', 'Hill', 71, 22),
 ('Mika', 'Häkkinen', 57, 20),
 ('Kimi', 'Räikkönen', 8, 20),
 ('Stirling', 'Moss', 475, 16),
 ('Jenson', 'Button', 18, 15),
 ('Graham', 'Hill', 289, 14),
 ('Jack', 'Brabham', 356, 14),
 ('Emerson', 'Fittipaldi', 224, 14),
 ('Alberto', 'Ascari', 647, 13),
 ('David', 'Coulthard', 14, 13),
 ('Mario', 'Andretti', 207, 12),
 ('Alan', 'Jones', 178, 12),
 ('Carlos', 'Reutemann', 199, 12),
 ('Jacques', 'Villeneuve', 35, 11),
 ('Felipe', 'Massa', 13, 11),
 ('Rubens', 'Barrichello', 22, 11),
 ('Ronnie', 'Peterson', 238, 10),
 ('Gerhard', 'Berger', 77, 10),
 ('Jody', 'Scheckter', 222, 10),
 ('James', 'Hunt', 231, 10),
 ('Mark', 'Webber', 17, 9),
 ('Denny', 'Hulme', 304, 8),
 ('Jacky', 'Ickx', 235, 8),
 ('Juan', 'Pablo Montoya', 31, 7),
 ('René', 'Arnoux', 163, 7),
 ('Gilles', 'Villeneuve', 203, 6),
 ('John', 'Surtees', 341, 6),
 ('Tony', 'Brooks', 479, 6),
 ('Ralf', 'Schumacher', 23, 6),
 ('Jochen', 'Rindt', 358, 6),
 ('Riccardo', 'Patrese', 119, 6),
 ('Jacques', 'Laffite', 172, 6),
 ('Michele', 'Alboreto', 105, 5),
 ('Daniel', 'Ricciardo', 817, 5),
 ('Nino', 'Farina', 642, 5),
 ('John', 'Watson', 187, 5),
 ('Keke', 'Rosberg', 177, 5),
 ('Clay', 'Regazzoni', 223, 5),
 ('Eddie', 'Irvine', 56, 4),
 ('Dan', 'Gurney', 364, 4),
 ('Bruce', 'McLaren', 360, 4),
 ('Johnny', 'Herbert', 65, 3),
 ('Didier', 'Pironi', 202, 3),
 ('Heinz-Harald', 'Frentzen', 49, 3),
 ('Phil', 'Hill', 403, 3),
 ('Mike', 'Hawthorn', 578, 3),
 ('Thierry', 'Boutsen', 123, 3),
 ('Peter', 'Collins', 581, 3),
 ('Giancarlo', 'Fisichella', 21, 3),
 ('Jean-Pierre', 'Jabouille', 219, 2),
 ('José Froilán', 'González', 498, 2),
 ('Wolfgang', 'von Trips', 476, 2),
 ('Bill', 'Vukovich', 657, 2),
 ('Pedro', 'Rodríguez', 345, 2),
 ('Maurice', 'Trintignant', 427, 2),
 ('Patrick', 'Tambay', 175, 2),
 ('Jo', 'Siffert', 346, 2),
 ('Valtteri', 'Bottas', 822, 2),
 ('Max', 'Verstappen', 830, 2),
 ('Elio', 'de Angelis', 173, 2),
 ('Patrick', 'Depailler', 221, 2),
 ('Peter', 'Revson', 309, 2),
 ('Peter', 'Gethin', 320, 1),
 ('Jo', 'Bonnier', 347, 1),
 ('Jim', 'Rathmann', 509, 1),
 ('Robert', 'Kubica', 9, 1),
 ('Luigi', 'Fagioli', 786, 1),
 ('Lorenzo', 'Bandini', 385, 1),
 ('Lee', 'Wallard', 766, 1),
 ('Heikki', 'Kovalainen', 5, 1),
 ('Gunnar', 'Nilsson', 255, 1),
 ('Jochen', 'Mass', 200, 1),
 ('Jean', 'Alesi', 55, 1),
 ('Jarno', 'Trulli', 15, 1),
 ('Jean-Pierre', 'Beltoise', 306, 1),
 ('Jimmy', 'Bryan', 525, 1),
 ('Ludovico', 'Scarfiotti', 375, 1),
 ('Luigi', 'Musso', 577, 1),
 ('Richie', 'Ginther', 386, 1),
 ('Bob', 'Sweikert', 628, 1),
 ('Troy', 'Ruttman', 526, 1),
 ('Innes', 'Ireland', 404, 1),
 ('Pastor', 'Maldonado', 813, 1),
 ('Johnnie', 'Parsons', 593, 1),
 ('Olivier', 'Panis', 44, 1),
 ('Rodger', 'Ward', 449, 1),
 ('Piero', 'Taruffi', 641, 1),
 ('François', 'Cevert', 327, 1),
 ('Giancarlo', 'Baghetti', 394, 1),
 ('Vittorio', 'Brambilla', 230, 1),
 ('Carlos', 'Pace', 250, 1),
 ('Sam', 'Hanks', 611, 1),
 ('Pat', 'Flaherty', 559, 1),
 ('Alessandro', 'Nannini', 145, 1)]

## Write a query that joins the tables results, constructors and drivers to show the amount of times a driver has driven a car from a constructor. In other words how many times one driver (for instance Schumacher) has driven a constructor (for instance Mercedes). Your table should have three fields: drivers.driverref, constructors.name and count(*) (showing the number of times the driver has driven the constructor).

In [6]:
%%sql
SELECT drivers.driverref, constructors.name, count(*) as "times_driven_for"
FROM results
INNER JOIN drivers ON
results.driverid = drivers.driverid
INNER JOIN constructors ON
results.constructorid = constructors.constructorid
GROUP BY drivers.driverid, constructors.name
ORDER BY count(*) DESC
limit 10

10 rows affected.


[('michael_schumacher', 'Ferrari', 181),
 ('coulthard', 'McLaren', 150),
 ('massa', 'Ferrari', 140),
 ('button', 'McLaren', 137),
 ('rosberg', 'Mercedes', 136),
 ('hakkinen', 'McLaren', 133),
 ('laffite', 'Ligier', 132),
 ('webber', 'Red Bull', 129),
 ('raikkonen', 'Ferrari', 126),
 ('vettel', 'Red Bull', 113)]

I have used "limit 10" but the query results in "2093 rows affected" without limit.

## Now we have found the most driven vehicles even for those who didn't finish the race. Extend the query from 2 by removing all tuples from the result where the status is not "Finished".

In [8]:
%%sql
SELECT drivers.driverref, constructors.name, count(*) as "times_driven_for"
FROM results
INNER JOIN drivers ON
results.driverid = drivers.driverid
INNER JOIN constructors ON
results.constructorid = constructors.constructorid
WHERE results.statusid = 1
GROUP BY drivers.driverid, constructors.name
ORDER BY count(*) DESC
limit 10

10 rows affected.


[('michael_schumacher', 'Ferrari', 141),
 ('massa', 'Ferrari', 112),
 ('rosberg', 'Mercedes', 109),
 ('webber', 'Red Bull', 100),
 ('vettel', 'Red Bull', 97),
 ('raikkonen', 'Ferrari', 92),
 ('alonso', 'Ferrari', 87),
 ('button', 'McLaren', 87),
 ('coulthard', 'McLaren', 85),
 ('hamilton', 'Mercedes', 85)]

I have used "limit 10" but the query results in "658 rows affected" without limit.

In [ ]:
%%sql CREATE MATERIALIZED VIEW times_driven_cache AS
SELECT drivers.driverid, drivers.driverref, constructors.name, count(*) as "times_driven_for"
FROM results
INNER JOIN drivers ON
results.driverid = drivers.driverid
INNER JOIN constructors ON
results.constructorid = constructors.constructorid
WHERE results.statusid = 1
GROUP BY drivers.driverid, constructors.name

## Create a third (and new) query that finds the amount of milliseconds spent in pitstop (see the pitstops relation) by each unique combinations of driverid and raceid.

In [10]:
%%sql
SELECT pitstops.driverid, results.constructorid, pitstops.raceid, sum(pitstops.milliseconds) as "time_spent_in_pitstop"
FROM results
INNER JOIN pitstops
ON results.driverid = pitstops.driverid AND results.raceid = pitstops.raceid
GROUP BY pitstops.driverid, pitstops.raceid, results.constructorid
ORDER BY pitstops.driverid
limit 10

10 rows affected.


[(1, 1, 841, Decimal('46426')),
 (1, 1, 842, Decimal('93011')),
 (1, 1, 843, Decimal('61978')),
 (1, 1, 844, Decimal('99637')),
 (1, 1, 845, Decimal('81457')),
 (1, 1, 846, Decimal('77769')),
 (1, 1, 848, Decimal('61587')),
 (1, 1, 849, Decimal('72505')),
 (1, 1, 850, Decimal('61881')),
 (1, 1, 851, Decimal('119200'))]

I have used "limit 10" but the query results in "2699 rows affected" without limit.

In [ ]:
%%sql CREATE MATERIALIZED VIEW pitstop_time_cache AS
SELECT pitstops.driverid, results.constructorid, pitstops.raceid, sum(pitstops.milliseconds) as "time_spent_in_pitstop"
FROM results
INNER JOIN pitstops
ON results.driverid = pitstops.driverid AND results.raceid = pitstops.raceid
GROUP BY pitstops.driverid, pitstops.raceid, results.constructorid
ORDER BY pitstops.driverid

## Time to put everything together. Use the query for 4 as a join subquery in query 3 (see slides on "a query within a query") to find the total amount of pitstop time, for each result in query 3

In [151]:
%%sql
SELECT ptc.driverid, tdc.driverref, tdc.name, sum(ptc.time_spent_in_pitstop) as "time_spent_in_pitstop"
from times_driven_cache as tdc, pitstop_time_cache as ptc
INNER JOIN constructors 
ON ptc.constructorid = constructors.constructorid
WHERE tdc.driverid = ptc.driverid AND constructors.name = tdc.name
GROUP BY tdc.name, ptc.driverid, tdc.driverref, ptc.constructorid
ORDER BY ptc.driverid

76 rows affected.


[(1, 'hamilton', 'McLaren', Decimal('2085508')),
 (1, 'hamilton', 'Mercedes', Decimal('12428578')),
 (2, 'heidfeld', 'Renault', Decimal('573333')),
 (3, 'rosberg', 'Mercedes', Decimal('12390412')),
 (4, 'alonso', 'Ferrari', Decimal('5012113')),
 (4, 'alonso', 'McLaren', Decimal('8217475')),
 (5, 'kovalainen', 'Caterham', Decimal('1281894')),
 (5, 'kovalainen', 'Lotus', Decimal('1047043')),
 (5, 'kovalainen', 'Lotus F1', Decimal('107321')),
 (8, 'raikkonen', 'Ferrari', Decimal('8343784')),
 (8, 'raikkonen', 'Lotus F1', Decimal('1657843')),
 (10, 'glock', 'Marussia', Decimal('935865')),
 (13, 'massa', 'Ferrari', Decimal('3376241')),
 (13, 'massa', 'Williams', Decimal('11874456')),
 (15, 'trulli', 'Lotus', Decimal('954047')),
 (16, 'sutil', 'Force India', Decimal('2152895')),
 (16, 'sutil', 'Sauber', Decimal('2038558')),
 (17, 'webber', 'Red Bull', Decimal('3345030')),
 (18, 'button', 'McLaren', Decimal('11800713')),
 (20, 'vettel', 'Ferrari', Decimal('9248461')),
 (20, 'vettel', 'Red Bull', Decimal('5170777')),
 (22, 'barrichello', 'Williams', Decimal('1230851')),
 (30, 'michael_schumacher', 'Mercedes', Decimal('2028748')),
 (37, 'rosa', 'HRT', Decimal('879844')),
 (37, 'rosa', 'Sauber', Decimal('54327')),
 (67, 'buemi', 'Toro Rosso', Decimal('984997')),
 (153, 'alguersuari', 'Toro Rosso', Decimal('1026081')),
 (154, 'grosjean', 'Haas F1 Team', Decimal('5107333')),
 (154, 'grosjean', 'Lotus F1', Decimal('4597738')),
 (155, 'kobayashi', 'Caterham', Decimal('1742177')),
 (155, 'kobayashi', 'Sauber', Decimal('1984793')),
 (807, 'hulkenberg', 'Force India', Decimal('10198930')),
 (807, 'hulkenberg', 'Renault', Decimal('2083151')),
 (807, 'hulkenberg', 'Sauber', Decimal('1066344')),
 (808, 'petrov', 'Caterham', Decimal('1058286')),
 (808, 'petrov', 'Renault', Decimal('1003106')),
 (811, 'bruno_senna', 'Renault', Decimal('538377')),
 (811, 'bruno_senna', 'Williams', Decimal('921875')),
 (813, 'maldonado', 'Lotus F1', Decimal('2867305')),
 (813, 'maldonado', 'Williams', Decimal('3029721')),
 (814, 'resta', 'Force India', Decimal('3036302')),
 (815, 'perez', 'Force India', Decimal('11876732')),
 (815, 'perez', 'McLaren', Decimal('963507')),
 (815, 'perez', 'Sauber', Decimal('1663803')),
 (816, 'ambrosio', 'Lotus F1', Decimal('21962')),
 (817, 'ricciardo', 'Red Bull', Decimal('11816929')),
 (817, 'ricciardo', 'Toro Rosso', Decimal('1955244')),
 (818, 'vergne', 'Toro Rosso', Decimal('4148771')),
 (819, 'pic', 'Caterham', Decimal('1004252')),
 (819, 'pic', 'Marussia', Decimal('926164')),
 (820, 'chilton', 'Marussia', Decimal('3212233')),
 (821, 'gutierrez', 'Haas F1 Team', Decimal('5567328')),
 (821, 'gutierrez', 'Sauber', Decimal('3223428')),
 (822, 'bottas', 'Mercedes', Decimal('1925619')),
 (822, 'bottas', 'Williams', Decimal('10665109')),
 (823, 'garde', 'Caterham', Decimal('1197610')),
 (824, 'jules_bianchi', 'Marussia', Decimal('2940201')),
 (825, 'kevin_magnussen', 'Haas F1 Team', Decimal('2032463')),
 (825, 'kevin_magnussen', 'McLaren', Decimal('2154571')),
 (825, 'kevin_magnussen', 'Renault', Decimal('5535336')),
 (826, 'kvyat', 'Red Bull', Decimal('1078624')),
 (826, 'kvyat', 'Toro Rosso', Decimal('8311267')),
 (828, 'ericsson', 'Caterham', Decimal('1843925')),
 (828, 'ericsson', 'Sauber', Decimal('5373074')),
 (830, 'max_verstappen', 'Red Bull', Decimal('5816742')),
 (830, 'max_verstappen', 'Toro Rosso', Decimal('2384382')),
 (831, 'nasr', 'Sauber', Decimal('8513533')),
 (832, 'sainz', 'Toro Rosso', Decimal('8356874')),
 (834, 'rossi', 'Manor Marussia', Decimal('311537')),
 (835, 'jolyon_palmer', 'Renault', Decimal('3584563')),
 (836, 'wehrlein', 'Manor Marussia', Decimal('5609314')),
 (836, 'wehrlein', 'Sauber', Decimal('2060905')),
 (838, 'vandoorne', 'McLaren', Decimal('2021671')),
 (839, 'ocon', 'Force India', Decimal('2105118')),
 (839, 'ocon', 'Manor Marussia', Decimal('5031572')),
 (840, 'stroll', 'Williams', Decimal('1890017'))]

To see that the query above is actually correct we can inspect the total time spent in pitstop by every driver with: 

```sql
SELECT driverid, sum(pitstop_time_cache.time_spent_in_pitstop) 
from pitstop_time_cache
GROUP BY driverid
```

For example Hamilton have spent a total time of 14514086 milliseconds in the pit.
2085508 + 12428578 = 14514086.

## As a last thing we want to find the driver with the least pistop time. However we want to control for the amount of races (see explanation below), so we need to include one last column: the average pitstop time per race. Your table should be sorted based on the ascending values of that average pitstop time.

In [152]:
%%sql
SELECT ptc.driverid, tdc.driverref, tdc.name, sum(ptc.time_spent_in_pitstop), 
round( sum(ptc.time_spent_in_pitstop) / count(ptc.driverid), 2) as "avg_pit_time"
from times_driven_cache as tdc, pitstop_time_cache as ptc
INNER JOIN constructors ON ptc.constructorid = constructors.constructorid
WHERE tdc.driverid = ptc.driverid AND constructors.name = tdc.name
GROUP BY tdc.name, ptc.driverid, tdc.driverref, ptc.constructorid
ORDER BY avg_pit_time ASC

76 rows affected.


[(816, 'ambrosio', 'Lotus F1', Decimal('21962'), Decimal('21962.00')),
 (8, 'raikkonen', 'Lotus F1', Decimal('1657843'), Decimal('46051.19')),
 (819, 'pic', 'Marussia', Decimal('926164'), Decimal('46308.20')),
 (815, 'perez', 'Sauber', Decimal('1663803'), Decimal('48935.38')),
 (10, 'glock', 'Marussia', Decimal('935865'), Decimal('49256.05')),
 (817, 'ricciardo', 'Toro Rosso', Decimal('1955244'), Decimal('50134.46')),
 (815, 'perez', 'McLaren', Decimal('963507'), Decimal('50710.89')),
 (811, 'bruno_senna', 'Williams', Decimal('921875'), Decimal('51215.28')),
 (155, 'kobayashi', 'Sauber', Decimal('1984793'), Decimal('52231.39')),
 (814, 'resta', 'Force India', Decimal('3036302'), Decimal('53268.46')),
 (30, 'michael_schumacher', 'Mercedes', Decimal('2028748'), Decimal('53388.11')),
 (5, 'kovalainen', 'Lotus F1', Decimal('107321'), Decimal('53660.50')),
 (826, 'kvyat', 'Red Bull', Decimal('1078624'), Decimal('53931.20')),
 (37, 'rosa', 'Sauber', Decimal('54327'), Decimal('54327.00')),
 (37, 'rosa', 'HRT', Decimal('879844'), Decimal('54990.25')),
 (808, 'petrov', 'Caterham', Decimal('1058286'), Decimal('55699.26')),
 (808, 'petrov', 'Renault', Decimal('1003106'), Decimal('55728.11')),
 (15, 'trulli', 'Lotus', Decimal('954047'), Decimal('56120.41')),
 (153, 'alguersuari', 'Toro Rosso', Decimal('1026081'), Decimal('57004.50')),
 (813, 'maldonado', 'Williams', Decimal('3029721'), Decimal('57164.55')),
 (2, 'heidfeld', 'Renault', Decimal('573333'), Decimal('57333.30')),
 (1, 'hamilton', 'McLaren', Decimal('2085508'), Decimal('57930.78')),
 (67, 'buemi', 'Toro Rosso', Decimal('984997'), Decimal('57941.00')),
 (5, 'kovalainen', 'Lotus', Decimal('1047043'), Decimal('58169.06')),
 (819, 'pic', 'Caterham', Decimal('1004252'), Decimal('59073.65')),
 (13, 'massa', 'Ferrari', Decimal('3376241'), Decimal('59232.30')),
 (807, 'hulkenberg', 'Sauber', Decimal('1066344'), Decimal('59241.33')),
 (17, 'webber', 'Red Bull', Decimal('3345030'), Decimal('59732.68')),
 (834, 'rossi', 'Manor Marussia', Decimal('311537'), Decimal('62307.40')),
 (16, 'sutil', 'Force India', Decimal('2152895'), Decimal('63320.44')),
 (5, 'kovalainen', 'Caterham', Decimal('1281894'), Decimal('64094.70')),
 (811, 'bruno_senna', 'Renault', Decimal('538377'), Decimal('67297.13')),
 (4, 'alonso', 'Ferrari', Decimal('5012113'), Decimal('67731.26')),
 (22, 'barrichello', 'Williams', Decimal('1230851'), Decimal('68380.61')),
 (20, 'vettel', 'Red Bull', Decimal('5170777'), Decimal('68943.69')),
 (154, 'grosjean', 'Lotus F1', Decimal('4597738'), Decimal('69662.70')),
 (823, 'garde', 'Caterham', Decimal('1197610'), Decimal('70447.65')),
 (818, 'vergne', 'Toro Rosso', Decimal('4148771'), Decimal('75432.20')),
 (821, 'gutierrez', 'Sauber', Decimal('3223428'), Decimal('89539.67')),
 (824, 'jules_bianchi', 'Marussia', Decimal('2940201'), Decimal('91881.28')),
 (820, 'chilton', 'Marussia', Decimal('3212233'), Decimal('97340.39')),
 (813, 'maldonado', 'Lotus F1', Decimal('2867305'), Decimal('102403.75')),
 (828, 'ericsson', 'Sauber', Decimal('5373074'), Decimal('105354.39')),
 (18, 'button', 'McLaren', Decimal('11800713'), Decimal('106312.73')),
 (830, 'max_verstappen', 'Toro Rosso', Decimal('2384382'), Decimal('108381.00')),
 (3, 'rosberg', 'Mercedes', Decimal('12390412'), Decimal('111625.33')),
 (825, 'kevin_magnussen', 'McLaren', Decimal('2154571'), Decimal('113398.47')),
 (16, 'sutil', 'Sauber', Decimal('2038558'), Decimal('119915.18')),
 (835, 'jolyon_palmer', 'Renault', Decimal('3584563'), Decimal('123605.62')),
 (8, 'raikkonen', 'Ferrari', Decimal('8343784'), Decimal('126420.97')),
 (822, 'bottas', 'Mercedes', Decimal('1925619'), Decimal('128374.60')),
 (840, 'stroll', 'Williams', Decimal('1890017'), Decimal('135001.21')),
 (1, 'hamilton', 'Mercedes', Decimal('12428578'), Decimal('136577.78')),
 (822, 'bottas', 'Williams', Decimal('10665109'), Decimal('140330.38')),
 (839, 'ocon', 'Force India', Decimal('2105118'), Decimal('140341.20')),
 (828, 'ericsson', 'Caterham', Decimal('18439